### DISCUSSION TITLE
Minimum recommended RAM ?

### DISCUSSION THREAD

**Discussion by nekhtan** [DependencyTrack/dependency-track#4124] [Q&A]
Hello,

According to the [official doc](https://docs.dependencytrack.org/getting-started/deploy-docker/) the absolute minimum amount of RAM is 4.5GB. But with 6GB it takes ~2 to 6 hours before DT uses all the available memory. This eventually leads to a server restart. 

![image](https://github.com/user-attachments/assets/3bbc7cd3-9d0d-467f-9ec3-2e5d0ece3687)
![image](https://github.com/user-attachments/assets/c9a482b2-b94f-4783-8e95-89560eeb45d8)
![image](https://github.com/user-attachments/assets/45a16094-3e64-4ccf-a0d0-6d47256a8461)

Here are some more information
* no load whatsoever, just the API server running
* roughly 10k "projects" stored in DT's database
* no peaks other than the memory (not so) slowly increasing
* I **think** it was much better in 4.11.4 (currently 4.11.7, and first noticed after upgrading to 4.11.5)

Do you know what could cause such behavior ?
Is the minimum RAM linked to the number of projects stored ? We naively guessed the RAM was mainly tied to ingestion processing, but maybe background tasks (to keep projects up-to-date & co) also need quite some resources. 

Anyway, we'll increase the memory and see if it helps.

Thanks in advance for your insight !

**Comment by nscuro** [comment]
> Do you know what could cause such behavior ? Is the minimum RAM linked to the number of projects stored ?

Yeah, asynchronous tasks will load projects (and their components) into memory, e.g. to perform vulnerability analysis on them. The fact that the majority of your heap size is from `G1 Old Gen` is a good giveaway that this might ([again](https://github.com/DependencyTrack/dependency-track/discussions/3818)) be linked to the ORM's L2 cache.

Unfortunately there's not much you can do outside of:

* Finding a way to force the garbage collector to clean up `G1 Old Gen` before it hits the memory limit
* Increasing the RAM for your Pod
* Disabling the L2 cache globally



Since (IIRC) you're running on k8s, can you determine if k8s kills the Pod with `OOMKilled`, or whether the container just quits with a non-zero exit code (i.e. the JVM kills itself)? The former would suggest that you can tweak the k8s config to avoid crashes.

> I think it was much better in 4.11.4 (currently 4.11.7, and first noticed after upgrading to 4.11.5)

I can't spot any changes between those versions that might impact this particular problem: https://github.com/DependencyTrack/dependency-track/compare/4.11.4...4.11.7. Obviously I don't want to discard your observation, but nothing jumps to my eye that would explain this [Verschlimmbesserung](https://en.wiktionary.org/wiki/verschlimmbessern).

---

I know it doesn't help you right now, but we have a Hyades instance with 40k projects where the memory usage graph looks like this:

![image](https://github.com/user-attachments/assets/fc62454e-a918-43cc-a139-22436481b4c9)

A big factor is that we were able to disable the L2 cache entirely. Unfortunately we can't do it as easily for everyone using v4 of DT :(

**Comment by nekhtan** [comment]
Hello @nscuro ,

Thanks for you answer !

> asynchronous tasks will load projects (and their components) into memory

But this is done sequentially, right ? So it has to be purgeable at some point or what do I miss ? Also should this kind of activity be visible somewhere in the provided dashboard ? Or do you know which log to track (e.g. ProjectMetricsUpdateTask) ?

> Finding a way to force the garbage collector to clean up G1 Old Gen before it hits the memory limit

I might try [-XX:InitiatingHeapOccupancyPercent](https://docs.oracle.com/cd/E40972_01/doc.70/e40973/cnf_jvmgc.htm#autoId2)

> Disabling the L2 cache globally

Since last time we completely disabled the L2 cache, so that can't be it.

> can you determine if k8s kills the Pod with OOMKilled, or whether the container just quits with a non-zero exit code (i.e. the JVM kills itself)?

I can't find any log of the crash so I would assume the JVM kills itself (?), to be confirmed.

> I can't spot any changes between those versions that might impact this particular problem [...]

That's what I thought as well, it might have been an unfortunate combination of factors that made me think that in the 1st place... But it was before the summer break so I can't remember anymore :D

Anyway, I can't wait for us to migrate to Hyades, it looks so promising !

Thanks again for your insight !

**Reply by nscuro** [reply]
> Since last time we completely disabled the L2 cache, so that can't be it.

Huh, that's *interesting*. I can't think of anything else that would stay around in old gen for this long. Then it *has* to be inefficient querying (i.e. loading ALL projects at once) I suspect.

Looking at your graphs, it seems as if memory keeps growing rapidly roughly 1 hour after the Pod restarts, is that correct?

If so, then the culprits have to be these two tasks, which both execute with a 1 hour delay after launch: 

https://github.com/DependencyTrack/dependency-track/blob/6c67c7bf50710129d5900d5efa15d8756a72cb3b/src/main/java/org/dependencytrack/tasks/TaskScheduler.java#L104-L108

And... well, looking at this code, I think I know who the prime subject is... :D

https://github.com/DependencyTrack/dependency-track/blob/6c67c7bf50710129d5900d5efa15d8756a72cb3b/src/main/java/org/dependencytrack/tasks/repositories/RepositoryMetaAnalyzerTask.java#L98-L109

Instead of fetching all projects and their components at once, it will need to do so in chunks of ~1000 to avoid excessive memory usage.

**Reply by nekhtan** [reply]
> Looking at your graphs, it seems as if memory keeps growing rapidly roughly 1 hour after the Pod restarts, is that correct?

Yes that looks like it 😲

> Instead of fetching all projects and their components at once, it will need to do so in chunks of ~1000 to avoid excessive memory usage.

That should do the trick yes, and maybe even make it configurable view UI/API ? Do you want me to create an issue for this ?

Anyway, for the time being we'll just drop the project after processing as we don't really need history and up-to-date data. If needed we'll process the SBOM all over again. Already tried that in a quick and dirty mode and it looks promising as the memory remains quite stable: between 500MB and 1.5GB.

**Reply by nscuro** [reply]
> and maybe even make it configurable view UI/API

I would delay that until we absolutely know it needs to be configurable.

> Anyway, for the time being we'll just drop the project after processing as we don't really need history and up-to-date data.

Alternatively, you can also mark projects as "inactive". Inactive projects are not analyzed on a schedule, and metrics are not refreshed for them. But you can still upload BOMs to them, and analysis will be performed then.

But yeah, if you don't need the history of course it's still best to simply delete the projects, as it will most definitely reduce storage and load on the database.

**Reply by nscuro** [reply]
Apologies this still wasn't fixed yet. Just raised https://github.com/DependencyTrack/dependency-track/pull/4311 and will include it in 4.12.1.
